aws/resource_aws_qldb_ledger.go

```diff
@@ -0,0 +1,286 @@
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/qldb"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceAwsQLDBLedger() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsQLDBLedgerCreate,
+		Read:   resourceAwsQLDBLedgerRead,
+		Update: resourceAwsQLDBLedgerUpdate,
+		Delete: resourceAwsQLDBLedgerDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		SchemaVersion: 1,
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 32),
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z0-9_-]+`), "must contain only alphanumeric characters, underscores, and hyphens"),
+				),
+			},
+
+			"permissions_mode": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{qldb.PermissionsModeAllowAll}, false),
+				Default:      qldb.PermissionsModeAllowAll, // Delete this line when AWS will support fetching permissions_mode
+			},
+
+			"deletion_protection": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsQLDBLedgerCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).qldbconn
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else {
+		name = resource.PrefixedUniqueId("tf")
+	}
+
+	if err := d.Set("name", name); err != nil {
+		return fmt.Errorf("error setting name: %s", err)
+	}
+
+	// Create the QLDB Ledger
+	createOpts := &qldb.CreateLedgerInput{
+		Name:               aws.String(d.Get("name").(string)),
+		PermissionsMode:    aws.String(d.Get("permissions_mode").(string)),
+		DeletionProtection: aws.Bool(d.Get("deletion_protection").(bool)),
+		Tags:               keyvaluetags.New(d.Get("tags").(map[string]interface{})).IgnoreAws().QldbTags(),
+	}
+
+	log.Printf("[DEBUG] QLDB Ledger create config: %#v", *createOpts)
+	qldbResp, err := conn.CreateLedger(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating QLDB Ledger: %s", err)
+	}
+
+	// Set QLDB ledger name
+	d.SetId(*qldbResp.Name)
+
+	log.Printf("[INFO] QLDB Ledger name: %s", d.Id())
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{qldb.LedgerStateCreating},
+		Target:     []string{qldb.LedgerStateActive},
+		Refresh:    qldbLedgerRefreshStatusFunc(conn, d.Id()),
+		Timeout:    8 * time.Minute,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		d.SetId("")
```

Comment:
We should not automatically remove resources in the `Create` function here -- the error may have just been a timeout waiting for it to become active, which means the resource should still be managed by Terraform.

```suggestion
```